QuestionAnswering/QA_Agent/deployment/fastapi_app/services/nlq2action/execute_action/sparql/process_query.py

```diff
@@ -0,0 +1,171 @@
+
+    def link_entities(self, sparql: str):
+        idx = 0
+        while idx < len(sparql):
+            # VALUES ?LandUseType { <LandUseType:\"residential\"> }
+            values_start = sparql.find("VALUES", idx)
+            if values_start < 0:
+                break
+
+            idx_start = values_start + len("VALUES") + 1
+            while idx_start < len(sparql) and sparql[idx_start].isspace():
+                idx_start += 1
+            if idx_start >= len(sparql):
+                break
+
+            # ?LandUseType { <LandUseType:\"residential\"> }
+            if sparql[idx_start] != "?":
+                break
+
+            idx = idx_start + 1
+            while idx < len(sparql) and (sparql[idx].isalnum() or sparql[idx] == "_"):
+                idx += 1
+            if idx >= len(sparql):
+                break
+
+            varnode = sparql[idx_start:idx]
+            idx_start = idx
+
+            while idx_start < len(sparql) and sparql[idx_start].isspace():
+                idx_start += 1
+            if idx_start >= len(sparql):
+                break
+
+            # { <LandUseType:\"residential\"> }
+            if sparql[idx_start] != "{":
+                break
+
+            tokens: List[str] = []
+
+            token_start = idx_start + 1
+            while True:
+                while token_start < len(sparql) and sparql[token_start].isspace():
+                    token_start += 1
+                if token_start >= len(sparql):
+                    break
+
+                if sparql[token_start] == "}":
+                    break
+
+                # assume no nested double quotations or pointed brackets
+                found = False
+                for c_start, c_end in ['""', "<>"]:
+                    if sparql[token_start] == c_start:
+                        token_end = sparql.find(c_end, token_start + 1)
+                        if token_end < 0:
+                            continue
+                        found = True
+                        tokens.append(sparql[token_start : token_end + 1])
+                        token_start = token_end + 1
+                        break
+
+                if not found:
+                    token_end = token_start + 1
+                    while token_end < len(sparql) and not sparql[token_end].isspace():
+                        token_end += 1
+                    if token_end >= len(sparql):
+                        break
+                    tokens.append(sparql[token_start : token_end + 1])
+                    token_start = token_end + 1
+
+            if sparql[token_start] != "}":
+                break
+            idx_end = token_start
+
+            values = []
+            for token in tokens:
+                values.extend(self._link_token(token))
+
+            # TODO: remove custom replacement of VALUES clause with FILTER
+            # once Karthik fixes the bug that causes ontop to throw error
+            # on VALUES of facility IRIs.
```

Comment:
What is the error you are seeing here?
I know that Ontop version 4 didn't handle `VALUES` efficiently but the stack now uses version 5 which does.

I also notice that you are working with IRIs of the form `<Type:...>` and I'm not sure that's a good idea. According to the [URI/IRI specifications](https://en.wikipedia.org/wiki/Uniform_Resource_Identifier#syntax) the first part of a URL/IRI (before the first ":") should be its schema, generally taken from the official list of schemas.

Reply:
This is the error that I'm seeing. I only encounter it when I use VALUES clause on IRIs of data centres and factories. I have asked @nagarajankarthik about this. I'm not sure what the timeline for fixing this issue is.
<img width="1014" alt="Screenshot 2024-05-07 at 8 49 58 PM" src="https://github.com/cambridge-cares/TheWorldAvatar/assets/124117301/6e47f6fc-d874-4afa-b6b2-0774869ce450">

For reference, here is the SPARQL query
```
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
PREFIX skos: <http://www.w3.org/2004/02/skos/core#>
PREFIX om: <http://www.ontology-of-units-of-measure.org/resource/om-2/>
PREFIX ontocompany: <http://www.theworldavatar.com/kg/ontocompany/>

SELECT DISTINCT ?IRI ?GeneratedHeatValue ?GeneratedHeatUnit WHERE {
  VALUES ?IRI { <https://www.theworldavatar.com/kg/infrastructure/datacenter/c84c1d32-69de-508f-a419-83096735aadd> <https://www.theworldavatar.com/kg/infrastructure/datacenter/7f8a8c48-516b-5757-a332-2a264052bdf8> }
  ?IRI ontocompany:hasGeneratedHeat/om:hasValue ?GeneratedHeat .
  ?GeneratedHeat om:hasNumericalValue ?GeneratedHeatValue .
  OPTIONAL { ?GeneratedHeat om:hasUnit/skos:notation ?GeneratedHeatUnit . }
}
```

Reply:
Regarding my use of `<Type:"surface form">`. The purpose of this is to serve as a marker for entities that need to be linked to exact IRIs and it is not meant to be a faithful representation of an URI/IRI as per general specifications. 

In literature, I have seen people propose special tokens for this purpose e.g. `<entity><type>Facility</type><label>ExxonMobil</label></entity>`. However, I find this too verbose. 

If you think that it is preferable to use special tokens in this case, I'll make the change.

Reply:
> Regarding my use of `<Type:"surface form">`. The purpose of this is to serve as a marker for entities that need to be linked to exact IRIs and it is not meant to be a faithful representation of an URI/IRI as per general specifications.

I don't fully understand the interplay between your code, Redis and the triples but what you say makes enough sense so its probably ok.